### PR TITLE
timeutil: fix scheduling on nth weekday of the month

### PR DESCRIPTION
**Pull request by bboozzoo**
The patch fixes scheduling on nth weekdays of the month. In particular checking
if the date being tested really is the nth occurrence of given weekday weekday
and whether if properly falls into range between nth weekdays.

Consider Feb 2017:
```
       February 2017
   Su Mo Tu We Th Fr Sa
             1  2  3  4
    5  6  7  8  9 10 11
   12 13 14 15 16 17 18
   19 20 21 22 23 24 25
   26 27 28
```
Schedule 'mon1-tue2' would incorrectly include dates 01-05.02, because they are
in the 1st week of the month, but not do not meet the schedule as they are not
'after' the first Monday of the month. Also dates 07-14.02 would be incorrectly
excluded as they are 'after' Tuesday in the 2nd week of the month, while in fact
the dates fall within the range of the schedule as they precede the 2nd Tuesday
of the month.

Address this by changing how the week-span checks are performed. In detail,
replace a simple week-count check with actually finding the nth occurrence of
given weekday in a month,

### CODE REVIEW

File: timeutil/schedule.go [hunk 1]
```diff
@@ -129,24 +129,38 @@ func (ws WeekSpan) String() string {
 	return ws.Start.String()
 }
 
+func findNthWeekDay(t time.Time, weekday time.Weekday, nthInMonth uint) time.Time {
+	// move to the beginning of the month
+	t = t.AddDate(0, 0, -t.Day()+1)
+
+	for nth := uint(0); t.Weekday() != weekday || nth != nthInMonth; {
+		t = t.Add(24 * time.Hour)
```
Comment (zyga) [inline timeutil/schedule.go]: Will this subtly break on leap year cases?

Comment (bboozzoo) [reply timeutil/schedule.go]: AFAICT by looking at `time`, the `Time.Add()` only deals with seconds. Denomalizing happens only when doing Day(), Weekday() and so on. As long as that part is correct we should be fine.

File: timeutil/schedule_test.go [hunk 2]
```diff
@@ -538,6 +538,35 @@ func (ts *timeutilSuite) TestScheduleNext(c *C) {
 			// expecting February, 27th, 10:00, 20 days and 23 hours
 			// from now
 			next: "503h-503h",
+		}, {
+			// from last Monday of the month to the Tuesday of
```
Comment (zyga) [inline timeutil/schedule_test.go]: From last Monday or from first Monday? The schedule below says `mon1` so I assume it's the first Monday of the month.

File: timeutil/schedule_test.go [hunk 3]
```diff
@@ -538,6 +538,35 @@ func (ts *timeutilSuite) TestScheduleNext(c *C) {
...
+			// the month, at 10:00
+			schedule: "mon1-tue2,10:00",
+			// Sunday, 22:00
+			last: "2017-02-06 10:00",
+			// first Monday of the month
+			now: "2017-02-07 11:00",
```
Comment (zyga) [inline timeutil/schedule_test.go]: 2017-02-07 is a Tuesday

File: timeutil/schedule_test.go [hunk 4]
```diff
@@ -538,6 +538,35 @@ func (ts *timeutilSuite) TestScheduleNext(c *C) {
 			// expecting February, 27th, 10:00, 20 days and 23 hours
 			// from now
 			next: "503h-503h",
+		}, {
+			// from last Monday of the month to the Tuesday of
+			// the month, at 10:00
+			schedule: "mon1-tue2,10:00",
+			// Sunday, 22:00
+			last: "2017-02-06 10:00",
```
Comment (zyga) [inline timeutil/schedule_test.go]: 2017-02-06 is a Monday.

**Pull request merged**